gcc/rust/checks/errors/rust-feature.cc

```diff
@@ -0,0 +1,67 @@
+// Copyright (C) 2020-2022 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#include "rust-feature.h"
+#include "rust-session-manager.h"
+
+namespace Rust {
+
+#define DECLARE_FEATURE(ST, N, RS, GS, I, E, D)                                \
+  {                                                                            \
+    ST, (N), (RS), (GS), (I), (E), (D)                                         \
+  }
+
+#define DECLARE_ACCEPTED_FEATURE(N, RS, GS, I, E, D)                           \
+  DECLARE_FEATURE (Feature::State::ACCEPTED, (N), (RS), (GS), (I), (E), (D))
+
+#define DECLARE_ACTIVE_FEATURE(N, RS, GS, I, E, D)                             \
+  DECLARE_FEATURE (Feature::State::ACTIVE, (N), (RS), (GS), (I), (E), (D))
+
+#define DECLARE_REMOVED_FEATURE(N, RS, GS, I, E, D)                            \
+  DECLARE_FEATURE (Feature::State::REMOVED, (N), (RS), (GS), (I), (E), (D))
+
+#define DECLARE_STABILIZED_FEATURE(N, RS, GS, I, E, D)                         \
+  DECLARE_FEATURE (Feature::State::STABILIZED, (N), (RS), (GS), (I), (E), (D))
+
+#define DECLARE_ACTIVE_FEATURE_IN_MAP(N, RS, GS, I, E, D)                      \
+  {                                                                            \
+    (N),                                                                       \
+    {                                                                          \
+      (Feature (DECLARE_ACCEPTED_FEATURE ((N), (RS), (GS), (I), (E), (D))))    \
+    }                                                                          \
+  }
+
+const std::map<std::string, const Feature &> Feature::feature_hash_map = {
+
+  DECLARE_ACTIVE_FEATURE_IN_MAP ("associated_type_bounds", "1.34.0", "0.0.0",
+				 52662,
+				 Optional<CompileOptions::Edition>::none (),
+				 ""),
+  DECLARE_ACTIVE_FEATURE_IN_MAP ("intrinsics", "1.0.0", "0.0.0", 0,
+				 Optional<CompileOptions::Edition>::none (),
+				 ""),
+  DECLARE_ACTIVE_FEATURE_IN_MAP ("rustc_attrs", "1.0.0", "0.0.0", 0,
+				 Optional<CompileOptions::Edition>::none (),
+				 ""),
+  DECLARE_ACTIVE_FEATURE_IN_MAP ("decl_macro", "1.0.0", "0.0.0", 0,
+				 Optional<CompileOptions::Edition>::none (),
+				 ""),
+
+};
```

Comment:
Is it not possible to avoid the use of macros here? I thought this would be okay with C++'s `constexpr` and so on